File: src/main/kotlin/com/n26/transactions/domain/TransactionRepository.java
package com.n26.transactions.domain;

import com.n26.transactions.statistics.domain.Statistics;
import kotlin.Pair;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Repository;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Repository
public class TransactionRepository {
    private List<Transaction> values = new ArrayList<>();

    public void addTransaction(Transaction request) {
        this.values.add(request);
    }

    public Statistics statisticsOfLast60Seconds() {
        final val validValues = values.stream().filter(it -> !it.expired()).collect(Collectors.toList());
        final int size = validValues.size();
        val count = size;
        return Statistics.Companion.of(count,
                pair("sum", sum(validValues, BigDecimal.ZERO)),
                pair("avg", avgOrDefault(validValues, BigDecimal.ZERO)),
                pair("max", maxOrDefault(validValues, BigDecimal.ZERO)),
                pair("min", minOrDefault(validValues, BigDecimal.ZERO)));
    }

    private kotlin.Pair<String, String> pair(String a, String b) {
        return new Pair<>(a, b);
    }

    private String sum(List<Transaction> values, BigDecimal defaultValue) {
        return formatted(sumAsBigDecimal(values, defaultValue));
    }

    @NotNull
    private BigDecimal sumAsBigDecimal(List<Transaction> values, BigDecimal defaultValue) {
        return values.stream().
                map(Transaction::getAmount)
                .reduce(BigDecimal::add)
                .orElse(defaultValue);
    }

    private String maxOrDefault(List<Transaction> values, BigDecimal defaultValue) {
        if (values.isEmpty()) {
            return formatted(defaultValue);
        } else {
            return formatted(values.stream().reduce(
                    (acc, ele) -> {
                        if (acc.getAmount().compareTo(ele.getAmount()) > 0) {
                            return acc;
                        } else {
                            return ele;
                        }
                    }).get().getAmount());
        }
    }

    private String minOrDefault(List<Transaction> values, BigDecimal defaultValue) {
        if (values.isEmpty()) {
            return formatted(defaultValue);
        } else {
            return formatted(values.stream().reduce(
                    (acc, ele) -> {
                        if (acc.getAmount().compareTo(ele.getAmount()) < 0) {
                            return acc;
                        } else {
                            return ele;
                        }
                    }).get().getAmount());
        }
    }

    private String avgOrDefault(List<Transaction> values, BigDecimal defaultValue) {
        if (values.isEmpty()) {
            return formatted(defaultValue);
        } else {
            final val amount = sumAsBigDecimal(values, defaultValue);
            BigDecimal total = new BigDecimal(values.size());
            return formatted(amount.divide(total, MathContext.DECIMAL128));
        }
    }

    private String formatted(BigDecimal t) {
        return t.setScale(2, RoundingMode.HALF_UP).toString();
    }

    public void deleteAllTransactions() {
        this.values.clear();
    }
}

File: src/main/kotlin/com/n26/transactions/TransactionRepository.kt
package com.n26.transactions

import com.n26.transactions.add.domain.AddTransaction
import com.n26.transactions.statistics.domain.Statistics
import org.springframework.stereotype.Repository
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

@Repository
open class TransactionRepository {
    private val values: MutableList<AddTransaction> = mutableListOf()

    open fun addTransaction(request: AddTransaction) {
        this.values.add(request)
    }

    open fun statisticsOfLast60Seconds(): Statistics {
        val validValues = values.filter { !it.expired() }
        val count = validValues.size
        return Statistics.of(count,
                "sum" to sum(validValues),
                "avg" to avgOrDefault(validValues, BigDecimal.ZERO),
                "max" to maxOrDefault(validValues, BigDecimal.ZERO),
                "min" to minOrDefault(validValues, BigDecimal.ZERO))
    }

    private fun sum(values: List<AddTransaction>): String = values.fold(BigDecimal.ZERO) { acc, it -> acc.add(it.amount) }.formatted()

    private fun maxOrDefault(values: List<AddTransaction>, defaultValue: BigDecimal): String {
        return if (values.isEmpty()) {
            defaultValue
        } else {
            values.reduce { acc, ele ->
                if (acc.amount > ele.amount) {
                    acc
                } else {
                    ele
                }
            }.amount
        }.formatted()
    }

    private fun minOrDefault(values: List<AddTransaction>, defaultValue: BigDecimal): String {
        return if (values.isEmpty()) {
            defaultValue
        } else {
            values.reduce { acc, ele ->
                if (acc.amount > ele.amount) {
                    ele
                } else {
                    acc
                }
            }.amount
        }.formatted()
    }

    private fun avgOrDefault(values: List<AddTransaction>, defaultValue: BigDecimal): String {
        val quantity = BigDecimal(values.size)
        val result = if (quantity == BigDecimal.ZERO) {
            defaultValue
        } else {
            val sum = values.fold(BigDecimal.ZERO) { acc, it -> acc.add(it.amount) }
            sum.divide(quantity, MathContext.DECIMAL128)
        }
        return result.formatted()
    }

    private fun BigDecimal.formatted(): String {
        return this.setScale(2, RoundingMode.HALF_UP).toString()
    }

    open fun deleteAllTransactions() {
        this.values.clear()
    }
}

